File: sudoku/CLang/Sudoku.c
#include <stdio.h>
#include <time.h>
int a[9][9];
int rows[9];
int cols[9];
int areas[9];
int cell_i[81];
int cell_j[81];
int cell_number = 0;
int best_candidate_threshold = 1;
int recursive_number = 0;


int getBestCandidates(int cell)
{
    int bestCandidates = 0;
    int bestCellIdx = -1;
    int bestCandidatesCount = 10;

    for (int idx = cell; idx < cell_number; idx++)
    {
        int i = cell_i[idx];
        int j = cell_j[idx];
        int candidates = 0;
        int candidatesCount = 0;

        for (int k=1; k<=9; k++)
        {
            int bits = 1<<k;
            if( ((rows[i] & bits)==0) && ((cols[j] & bits)==0) && ((areas[3*(i/3)+j/3] & bits)==0))
            {
                candidates |= bits;
                candidatesCount +=1;
            }
        }

        if (candidatesCount <= bestCandidatesCount)
        {
            bestCandidates = candidates;
            bestCellIdx = idx;
            bestCandidatesCount = candidatesCount;
            if (bestCandidatesCount < best_candidate_threshold)
                break;
        }
    }

    int tmpI = cell_i[cell];
    int tmpJ = cell_j[cell];
    cell_i[cell] = cell_i[bestCellIdx];
    cell_j[cell] = cell_j[bestCellIdx];
    cell_i[bestCellIdx] = tmpI;
    cell_j[bestCellIdx] = tmpJ;

    return bestCandidates;
}

void flag(int i, int j, int bitPos)
{
    int bits = 1<<bitPos;
    rows[i] |= bits;
    cols[j] |= bits;
    areas[3*(i/3) + j/3] |= bits;
}

void unflag(int i, int j, int bitPos)
{
    int bits = ~(1<<bitPos);
    rows[i] &= bits;
    cols[j] &= bits;
    areas[3*(i/3) + j/3] &= bits;
}

void print_result(int matrix[9][9])
{
    for(int i = 0; i<9; i++)
    {
        for (int j = 0; j<9; j++)
        {
            printf("%3d",matrix[i][j]);
        }
        printf("\n");
    }
}

int solve(int cell)
{
    recursive_number +=1;
    if (cell > cell_number - 1)
        return 1;
    int candidates = getBestCandidates(cell);
    int i = cell_i[cell];
    int j = cell_j[cell];
    for (int v= 1; v<=9; v++)
    {
        if ((candidates & (1<<v)))
        {
            a[i][j] = v;
            flag(i, j, v);
            if (solve(cell+1) == 1)
                return 1;
            unflag(i, j, v);
        }
    }

    a[i][j] = 0;
    return 0;
}

void sudokuRun()
{
    solve(0);
}

void init()
{
    int example[9][9] = {
        {0,0,0,0,0,6,0,0,0},
        {0,5,9,0,0,0,0,0,8},
        {2,0,0,0,0,8,0,0,0},
        {0,4,5,0,0,0,0,0,0},
        {0,0,3,0,0,0,0,0,0},
        {0,0,6,0,0,3,0,5,4},
        {0,0,0,3,2,5,0,0,6},
        {0,0,0,0,0,0,0,0,0},
        {0,0,0,0,0,0,0,0,0}
        };

    for (int i = 0; i<9; i++)
    {
        for(int j = 0; j<9; j++)
        {
            a[i][j] = example[i][j];
            if(a[i][j]==0)
            {
                cell_i[cell_number] = i;
                cell_j[cell_number] = j;
                cell_number +=1;
            }
            else
            {
                int bits = 1<<a[i][j];
                rows[i] |= bits;
                cols[j] |= bits;
                areas[3*(i/3)+j/3] |= bits;
            }
        }
    }
}

int main()
{
    init();

    print_result(a);

    printf("-------------\n");
    
    clock_t t1, t2;  
    t1 = clock(); 

    sudokuRun();
    
    t2 = clock();   

    float diff = ((float)(t2 - t1) / 1000000.0F ) * 1000;   
    

    print_result(a);

    printf("%f %d \n",diff, recursive_number);  
    
    return 0;
}

File: sudoku/sudoku.py
import time

class Sudoku:
    missed_cells = []
    cell_number = 0
    count = 0
    rows = []
    cols = []
    areas = []

    def __init__(self, a):
        self.a = a
        self.rows = [0 for x in range(9)]
        self.cols = [0 for x in range(9)]
        self.areas = [0 for x in range(9)]

        for i in range(9):
            for j in range(9):
                if a[i][j] == 0:
                    self.missed_cells.append([i, j])
                else:
                    bits = 1<<a[i][j]
                    self.rows[i] = self.rows[i] | bits
                    self.cols[j] = self.cols[j] | bits
                    self.areas[3*int(i/3)+int(j/3)] = self.areas[3*int(i/3)+int(j/3)] | bits

        self.cell_number = len(self.missed_cells)
        self.best_candidate_threshold = 1

    def run(self):
        self.count = 0
        self.solve(0)
        return self.a

    def solve(self, cell_i):
        self.count += 1
        if cell_i > self.cell_number - 1:
            return True
        candidates = self.get_best_cell(cell_i)
        cell = self.missed_cells[cell_i]
        i = cell[0]
        j = cell[1]
        for v in range(1,10):
            if candidates & (1<<v):
                self.a[i][j] = v
                self.flag(i,j,v)
                
                if self.solve(cell_i + 1):
                    return True
                self.unflag(i,j,v)

        self.a[i][j] = 0
        return False

    def get_best_cell(self, cell_i):
        if cell_i > self.cell_number - 1:
            return None, []
        best_candidates = 0
        best_cell_idx = -1
        best_candidates_count = 10
        for idx in range(cell_i, self.cell_number):
            cell = self.missed_cells[idx]
            i = cell[0]
            j = cell[1]
            candidates = 0
            candidates_count = 0
            
            for k in range(1,10):
                bits = 1<<k
                if (not self.rows[i] & bits) and (not self.cols[j] & bits) and (not self.areas[3*int(i/3)+int(j/3)] & bits):
                    candidates = candidates | bits
                    candidates_count += 1
            if candidates_count <= best_candidates_count:
                best_candidates = candidates
                best_cell_idx = idx
                best_candidates_count = candidates_count
                if candidates_count <= self.best_candidate_threshold:
                    break

        tmp = self.missed_cells[cell_i]
        self.missed_cells[cell_i] = self.missed_cells[best_cell_idx]
        self.missed_cells[best_cell_idx] = tmp
        
        return best_candidates
    
    def flag(self, i, j, bitPos):
        bits = 1<<bitPos
        self.rows[i] = self.rows[i] | bits
        self.cols[j] = self.cols[j] | bits
        self.areas[3*int(i/3)+int(j/3)] = self.areas[3*int(i/3)+int(j/3)] | bits
    
    def unflag(self, i, j, bitPos):
        bits = ~(1<<bitPos)
        self.rows[i] = self.rows[i] & bits
        self.cols[j] = self.cols[j] & bits
        self.areas[3*int(i/3)+int(j/3)] = self.areas[3*int(i/3)+int(j/3)] & bits

def main():
    from example import hard1
    from example import hard2
    from example import easy
    from example import super_hard
    from example import norvig1
    from example import norvig2
    from example import norvig3 # very hard, still can not be solved

    a = super_hard

    print_result(a)
    print("----------------")
    sudoku = Sudoku(a)

    start = time.clock()

    result = sudoku.run()

    end = time.clock()

    print(end-start, sudoku.count)
    print_result(result)

def print_result(result):
    for i in range(9):
        print(result[i])

if __name__ == "__main__":
    main()